### PR TITLE
Make the development version identifier PEP 440 compliant.

### LINKED ISSUE #8953: SymPy has invalid version numbers for PEP 440
In particular the version number for the dev versions, e.g. `0.7.6-git` is invalid. The latest version of setuptools is now treating `0.7.6-git` to be older than `0.7.6`. This is bad. For example it is breaking upstream tests in PyDy for SymPy master: https://github.com/pydy/pydy/pull/113.

### PR DESCRIPTION
**Pull request by moorepants**
Fixes #8953.

### CODE REVIEW

**Comment by moorepants** [comment]
@asmeurer What do you think of this?

File: sympy/interactive/session.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def _make_message(ipython=True, quiet=False, source=None):
                 _source += '>>> ' + line + '\n'
 
         doc_version = sympy_version
-        if doc_version.find('-git') >= 0:
+        if doc_version.find('-git') >= 0 or doc_version.find('dev') >= 0:
```
Comment (certik) [inline sympy/interactive/session.py]: Do we need to keep the `doc_version.find('-git') >= 0` part in there?

**Comment by certik** [comment]
I am +1 to the change, as we need to be PEP 440 compliant.

File: sympy/interactive/session.py (hunk 1 above)
Comment (moorepants) [inline sympy/interactive/session.py]: I guess not, but I just left it there in case. I didn't really read what all that code does. I'll look at it.

**Comment by asmeurer** [comment]
Is the 1 (at the end of `.dev1`) needed?

**Comment by moorepants** [comment]
The spec seems to require it but I haven't tried it without.

**Comment by moorepants** [comment]
@asmeurer Looks like it works without the numeral at the end:
```
In [6]: from pkg_resources import parse_version

In [7]: parse_version('0.7.7.dev') > parse_version('0.7.6')
Out[7]: True
```
But I can't find that mentioned in the PEP.

**Comment by moorepants** [comment]
Quote for PEP 440:

> The developmental release segment consists of the string .dev , followed by a non-negative integer value.

But 0 is allowed. I'm not sure if no number is equivalent to 0. Seems like setuptools thinks so:

```
In [1]: from pkg_resources import parse_version

In [2]: parse_version('0.7.6.dev') == parse_version('0.7.6.dev0')
Out[2]: True
```

Ahah!

```
Implicit development release number
=============================

Development releases allow omiting the numeral in which case it is implicitly assumed to be 0 . The normal form for this is to include the 0 explicitly. This allows versions such as 1.2.dev which is normalized to 1.2.dev0 .
```

I'll remove the zero. It is fine.

**Comment by moorepants** [comment]
I'm not sure what to do about this line:
https://github.com/sympy/sympy/blob/master/release/fabfile.py#L638

There are these prefixes `sympy-`, what is that all about?

**Comment by moorepants** [comment]
@asmeurer @certik I think this is ready to go in.

**Comment by moorepants** [comment]
Should I squash?

**Comment by certik** [comment]
I am fine to merge as it is. @asmeurer are you +1 to this PR?

**Comment by moorepants** [comment]
@asmeurer Can I get a +1 on this?

**Comment by moorepants** [comment]
@asmeurer I'd like to merge this soon. Are you fine with it?

**Comment by asmeurer** [comment]
Is the 0.7.6.rc1 -> 0.7.6rc1 change als required?

**Comment by moorepants** [comment]
The is the desired form: [N!]N(.N)*[{a|b|rc}N][.postN][.devN]

No period is preferred: https://www.python.org/dev/peps/pep-0440/#pre-releases

But there is an exception: https://www.python.org/dev/peps/pep-0440/#pre-release-separators

So a period form `0.7.6.rc1` is permissible.

**Comment by moorepants** [comment]
I'm still interested in merging this. @asmeurer What do you prefer about the rc versions?

**Comment by moorepants** [comment]
@asmeurer I'd really like to get this merged. Can you tell me what you want me to do with this? This cause downstream if you test against development versions (i.e. PyDy).

**Comment by certik** [comment]
I am ok with the PR. Aaron,  any objections?

**Comment by moorepants** [comment]
@asmeurer How do you want the rc versions to work: pep 440 recommendation or what you had (an exception to pep 440).

**Comment by moorepants** [comment]
@asmeurer Ping on this one again. I'm happy to fix it any way you prefer, just let me know.

**Comment by certik** [comment]
This is fine with me.

**Comment by moorepants** [comment]
I'm merging this. I just made the RC versions compliant with PEP 440 and left them as they were previously given the exception. That should resolve @asmeurer's question above.

**Comment by certik** [comment]
You might have been too fast merging something you didn't want to merge.

**Comment by certik** [comment]
Next time wait for a review. I am sorry that @asmeurer was too slow to respond. Review is for catching exactly the type of mistakes like this one.

**Comment by moorepants** [comment]
Sorry, but I've been waiting 3 months for the review. Little too impatient at this point. I'll try to chill out in the future.

We need to figure out how to get reviewing to be more efficient, as it is easy to get frustrated, impatient, or tired of waiting and just abandon it.

Questions that come to mind:

1. Are requirements to strict for merging code?
2. Is there any way to minimize bike-shedding?
3. Is the only solution to boost the number of reviewers?
4. How to boost # of reviewers?
5. For huge PR's how to effectively review?

I think about academic paper reviews. They consume lots of peoples time but most people (I think) just spend like 1-2 hr's per paper (maybe even less on average) looking for glaringly deficienes. Code review on the other seems focus more on details. It is easier to complain about a PEP8 "violation" than step back and think about the big picture of the code submission.

Anyway, we need to figure out how to reduce our massive issue and PR queue.

**Comment by moorepants** [comment]
We should collect some data from our repo to see some things like:

- average # days from PR submission to merge
- average # of reviewers
- who's doing the bulk of the reviewing, i.e. lots of reviewers or only a handful?

It would be interesting to know some stats.

**Comment by certik** [comment]
Well, this was an important change that I wanted @asmeurer approval. I think he is quite busy. I think we need to figure out how to scale the review process better.

One idea is to let our GSoC students review PRs as part of their job. It's a great way for them to get involved as well as learn new things and meet more people. And PRs get reviewed, then we just need somebody with push access to actually merge, if he or she agrees with the review (i.e. like a journal editor).

**Comment by moorepants** [comment]
> Well, this was an important change that I wanted @asmeurer approval. I think he is quite busy. I think we need to figure out how to scale the review process better.

Sure, I agree. In this case, I thought this was a very important change, thus labeling it "High Priority" three months ago. I was breaking the downstream PyDy build system we use to ensure PyDy is always working with SymPy master. How do we ensure that bugs like this are reviewed in a timely fashion?

> One idea is to let our GSoC students review PRs as part of their job.

Absolutely, if the GSoC funds only helped reduced our PR queue down to < 20, then maybe that's worth more than some new feature. I mandated PR review for the three PyDy students last year and it helped a lot. I think requiring the students to review PRs is more valuable than having them write a blog post, by far.

How many +1's are necessary? In this PR's case, you +1 multiple times and Aaron never -1 or anything. He just seemed too busy to respond.

This is interesting: http://blog.codacy.com/2015/04/28/visualizing-programming-languages-and-pull-request-lifespan/ Note that merging seems to take longer in Python repos for some reason.

**Comment by toolforger** [comment]
Am 01.05.2015 um 00:49 schrieb Jason K. Moore:
> 1. Are requirements to strict for merging code?

Actually I feel we're sometimes letting in too low-quality code.
"Too strict" maybe in a bike-shedding sense: easily checkable stuff gets 
pointed out all the time. OTOH it doesn't hurt to have them fixed right 
off the bat. When I'm in such a small-things commit cycle, my biggest 
worry is that it will require another round of review and risks being 
stalled; I'd really prefer if things got fixed.

> 2. Is there any way to minimize bike-shedding?

Have people state how important some issue is. E.g. I routinely state 
"you could improve so-and-so, but it's not important and not going to be 
-1 if other things are more important".

I'd like to see an easy way for the reviewer to fix the PR.
Right now, the reviewer would have to check out the PR and send a PR on 
the PR. I think the overall process is a bit too complicated: The 
reviewer must know how to pull the PR, set up a writable repo to push it 
back to GH, and the reviewee must know how to merge and deal with the 
occasional merge conflict.
It's far easier to add a comment on the PR and tell the reviewee to fix it.
*Maybe* we should tell contributors to make their repositories 
world-writable (we'd have to work around the fact that GitHub requires 
explicitly adding collaborators to a repo). Then reviewers could simply 
apply their bikeshedding, all the reviewee would have to do would be to 
simply merge from GitHub to their local copy, or push --force if they do 
not wish to apply the edits.

> 3. Is the only solution to boost the number of reviewers?

Boost number of reviews actually done.
Adding more reviewers isn't easy. Reviewing is a skill that needs time 
to develop.

> 4. How to boost # of reviewers?  (-> boost number of reviews done)

Contributors could be tasked with "before a Contributor PR can be 
merged/closed, two non-Contributor ones must be merged/closed". Or 
whatever the contributor/non-contributor merge ratio is. ("Contributor" 
means "people who can merge".)
If that's too much work, Contributors will start either urging people to 
review or urging Owners to appoint more Contributors.

> 5. For huge PR's how to effectively review?

Ask the requester to split it.

> Anyway, we need to figure out how to reduce our massive issue and PR queue.

I couldn't agree more.

**Comment by toolforger** [comment]
Am 01.05.2015 um 01:03 schrieb Jason K. Moore:
>> One idea is to let our GSoC students review PRs as part of their job.
>
> Absolutely, if the GSoC funds only helped reduced our PR queue down to < 20, then maybe that's worth more than some new feature.

+1

 > I mandated PR review for the three PyDy students last year and it 
helped a lot. I think requiring the students to review PRs is more 
valuable than having them write a blog post, by far.

Actually that's advice from the leads of successful projects, so I'd be 
+1 to try that.

> This is interesting: http://blog.codacy.com/2015/04/28/visualizing-programming-languages-and-pull-request-lifespan/ Note that merging seems to take longer in Python repos for some reason.

I don't know the average merge time for SymPy, but it's at least a 
magnitude above the Python average of less than three days. So any issue 
with Python itself isn't what should bother us.

**Pull request merged**